gandy/tests/test_models/test_gans.py

```diff
@@ -0,0 +1,200 @@
+        create batches of boostrapped data.
+        """
+        # check NOT one hot encoded Ys
+        Xs = 'Xs'
+        Ys = 'Ys'
+        subject = gans.GAN(xshape=(4,), yshape=(2,), n_classes=10)
+        subject.generate_data = mock.MagicMock(
+            name='generate_data', return_value=('classes', 'points'))
+        kwargs = dict(bacthes=1, batch_size=5)
+        with mock.patch('deepchem.metrics.to_one_hot',
+                        return_value='one_hot_classes') as mocked_one_hot:
+            result = list(subject.iterbacthes(Xs, Ys, kwargs))
+            subject.generate_data.assert_called_with(Xs, Ys, 5)
+            expected_result = {subject._model.data_inputs[0]: 'points',
+                               subject._model.conditional_inputs[0]:
+                               'classes'}
+            self.assertEqual(expected_result, result)
+        # check one hot encoded Ys
+        Xs = 'Xs'
+        Ys = [[0, 1], [1, 0], [1, 0]]
+        subject = gans.GAN(xshape=(4,), yshape=(2,), n_classes=10)
+        subject.generate_data = mock.MagicMock(
+            name='generate_data', return_value=('classes', 'points'))
+        kwargs = dict(bacthes=1, batch_size=5)
+        with mock.patch('deepchem.metrics.to_one_hot',
+                        return_value='one_hot_classes') as mocked_one_hot:
+            result = list(subject.iterbacthes(Xs, Ys, kwargs))
+            subject.generate_data.assert_called_with(Xs, Ys, 5)
+            mocked_one_hot.assert_called_with('classes', 10)
+            expected_result = {subject._model.data_inputs[0]: 'points',
+                               subject._model.conditional_inputs[0]:
+                               'one_hot_classes'}
+            self.assertEqual(expected_result, result)
+        return
+
+    @unittest.mock.patch('gandy.models.gans.GAN._build', return_value='Model')
+    def test_generate_data(self, mocked__build):
+        """
+        Test generate_data function.
+
+        The generate_data function creates batches of boostrapped data.
+        """
+        Xs = ['x1', 'x2', 'x3']
+        Ys = ['y1', 'y2', 'y3']
+        subject = gans.GAN(xshape=(4,), yshape=(2,), n_classes=1)
+        classes, points = subject.generate_data(Xs, Ys, 5)
+        self.assertEqual(len(classes), 5)
+        self.assertEqual(len(points), 5)
+        return
+
+    @unittest.mock.patch('gandy.models.gans.GAN._build', return_value='Model')
+    def test__save(self, mocked__build):
+        """
+        Test save function.
+
+        This checks that a file is written with the appropriate name.
+        """
+        # test path save
+        subject = gans.GAN(xshape=(4,), yshape=(2,), n_classes=10)
+        subject._model.save = mock.MagicMock('save')
+        subject.save('path')
+        subject._model.save.assert_called_with('path')
+        # test h5 save
+        subject.save('test_model.h5')
+        subject._model.save.assert_called_with('test_model.h5')
```

Comment:
I think the second test is unnecessary? if it passes the first one it will pass the second one. Unless `subject.save` conditionally changes the input eg. if h5, "test_model.h5" becomes "test_model" before passing to `_model.save`, you wouldn't need a second